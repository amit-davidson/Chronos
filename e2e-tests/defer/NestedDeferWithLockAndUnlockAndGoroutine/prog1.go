package main

import (
	"math/rand"
	"sync"
)

var a int
var cond = false

func main() {
	mutex := sync.Mutex{}
	if rand.Int() > 0 {
		defer mutex.Lock()
	}
	defer func() {
		mutex.Lock()
		a = 6
		defer mutex.Unlock()
	}()
	mutex.Lock()
	defer func() {
		go func() {
			a = 7
			defer mutex.Unlock()
		}()
	}()
}
