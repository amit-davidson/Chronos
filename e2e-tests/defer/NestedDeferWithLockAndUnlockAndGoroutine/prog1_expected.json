{
	"Lockset": {
		"ExistingLocks": {},
		"ExistingUnlocks": {
			"github.com/amit-davidson/Chronos/e2e-tests/defer/NestedDeferWithLockAndUnlockAndGoroutine/prog1.go:12:2": "github.com/amit-davidson/Chronos/e2e-tests/defer/NestedDeferWithLockAndUnlockAndGoroutine/prog1.go:19:21"
		}
	},
	"GuardedAccess": [
		{
			"ID": 1,
			"Pos": "github.com/amit-davidson/Chronos/e2e-tests/defer/NestedDeferWithLockAndUnlockAndGoroutine/prog1.go:13:16",
			"Value": "github.com/amit-davidson/Chronos/e2e-tests/defer/NestedDeferWithLockAndUnlockAndGoroutine/prog1.go:13:13",
			"OpKind": 0,
			"LocksetJson": {
				"ExistingLocks": {},
				"ExistingUnlocks": {}
			},
			"State": {
				"GoroutineID": 1,
				"Clock": {
					"1": 1
				}
			}
		},
		{
			"ID": 2,
			"Pos": "github.com/amit-davidson/Chronos/e2e-tests/defer/NestedDeferWithLockAndUnlockAndGoroutine/prog1.go:13:16",
			"Value": "",
			"OpKind": 0,
			"LocksetJson": {
				"ExistingLocks": {},
				"ExistingUnlocks": {}
			},
			"State": {
				"GoroutineID": 1,
				"Clock": {
					"1": 2
				}
			}
		},
		{
			"ID": 3,
			"Pos": "",
			"Value": "github.com/amit-davidson/Chronos/e2e-tests/defer/NestedDeferWithLockAndUnlockAndGoroutine/prog1.go:13:16",
			"OpKind": 0,
			"LocksetJson": {
				"ExistingLocks": {},
				"ExistingUnlocks": {}
			},
			"State": {
				"GoroutineID": 1,
				"Clock": {
					"1": 3
				}
			}
		},
		{
			"ID": 4,
			"Pos": "github.com/amit-davidson/Chronos/e2e-tests/defer/NestedDeferWithLockAndUnlockAndGoroutine/prog1.go:24:4",
			"Value": "",
			"OpKind": 0,
			"LocksetJson": {
				"ExistingLocks": {},
				"ExistingUnlocks": {}
			},
			"State": {
				"GoroutineID": 2,
				"Clock": {
					"1": 4,
					"2": 1
				}
			}
		},
		{
			"ID": 5,
			"Pos": "github.com/amit-davidson/Chronos/e2e-tests/defer/NestedDeferWithLockAndUnlockAndGoroutine/prog1.go:24:4",
			"Value": "github.com/amit-davidson/Chronos/e2e-tests/defer/NestedDeferWithLockAndUnlockAndGoroutine/prog1.go:8:5",
			"OpKind": 1,
			"LocksetJson": {
				"ExistingLocks": {},
				"ExistingUnlocks": {}
			},
			"State": {
				"GoroutineID": 2,
				"Clock": {
					"1": 4,
					"2": 2
				}
			}
		},
		{
			"ID": 6,
			"Pos": "github.com/amit-davidson/Chronos/e2e-tests/defer/NestedDeferWithLockAndUnlockAndGoroutine/prog1.go:18:3",
			"Value": "",
			"OpKind": 0,
			"LocksetJson": {
				"ExistingLocks": {
					"github.com/amit-davidson/Chronos/e2e-tests/defer/NestedDeferWithLockAndUnlockAndGoroutine/prog1.go:12:2": "github.com/amit-davidson/Chronos/e2e-tests/defer/NestedDeferWithLockAndUnlockAndGoroutine/prog1.go:17:13"
				},
				"ExistingUnlocks": {}
			},
			"State": {
				"GoroutineID": 1,
				"Clock": {
					"1": 5
				}
			}
		},
		{
			"ID": 7,
			"Pos": "github.com/amit-davidson/Chronos/e2e-tests/defer/NestedDeferWithLockAndUnlockAndGoroutine/prog1.go:18:3",
			"Value": "github.com/amit-davidson/Chronos/e2e-tests/defer/NestedDeferWithLockAndUnlockAndGoroutine/prog1.go:8:5",
			"OpKind": 1,
			"LocksetJson": {
				"ExistingLocks": {
					"github.com/amit-davidson/Chronos/e2e-tests/defer/NestedDeferWithLockAndUnlockAndGoroutine/prog1.go:12:2": "github.com/amit-davidson/Chronos/e2e-tests/defer/NestedDeferWithLockAndUnlockAndGoroutine/prog1.go:17:13"
				},
				"ExistingUnlocks": {}
			},
			"State": {
				"GoroutineID": 1,
				"Clock": {
					"1": 6
				}
			}
		}
	]
}